### Note
* Conclusions at the bottom of the file.

In [264]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [265]:
players = len(purchase_data['SN'].value_counts())
total_players = pd.DataFrame({'Total Players':players},index=[0]) #forced to specify an index since using scalar values
total_players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [266]:
items = len(purchase_data['Item ID'].value_counts())
avg_price = purchase_data['Price'].mean()
purchases = purchase_data['Purchase ID'].count()
revenue = purchase_data['Price'].sum()

#create dataframe
purchase_summary = pd.DataFrame({'Number of Unique Items':[items],
                                'Average Price':[avg_price],
                                'Number of Purchases':[purchases],
                                'Total Revenue':[revenue]})
#reformat dataframe
purchase_summary['Average Price'] = purchase_summary['Average Price'].map('${0:,.2f}'.format)
purchase_summary['Total Revenue'] = purchase_summary['Total Revenue'].map('${0:,.2f}'.format)

purchase_summary

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [267]:
purchase_data['SN'].value_counts() #looks like some users made more than one purchase (remember 576 players)
players_df = purchase_data.drop_duplicates(subset='SN') #take the first time a user shows up to avoid counting them again
gender_count = players_df['Gender'].value_counts() #this is a series
genders = gender_count.index
percents = gender_count/players #looks like you can divide a series by an int (nice)

#create and reformat a dataframe containing the desired output
gender_summary = pd.DataFrame({'Total Count':counts,
                              'Percentage of Players':percents}, 
                              index=genders)
gender_summary['Percentage of Players'] = gender_summary['Percentage of Players'].map('{0:,.2%}'.format)
gender_summary

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [268]:
purchases = purchase_data['Gender'].value_counts()
grouped_df = purchase_data.groupby(['Gender']) 
average_purchase_price = grouped_df['Price'].mean()
total_purchase = grouped_df['Price'].sum() #this calculates total purchase value by each gender
average_total_purchase = total_purchase/gender_count #remember gender_count from previous section is for unique players

#create and format summary dataframe
gender_purchases_summary = pd.DataFrame({'Purchase Count':purchases,
                                        'Average Purchase Price':average_purchase_price,
                                        'Total Purchase Value':total_purchase,
                                        'Avg Total Purchase per Person':average_total_purchase},
                                       index=genders) #genders defined in the section above
gender_purchases_summary['Average Purchase Price'] = gender_purchases_summary['Average Purchase Price'].map('${0:,.2f}'.format)
gender_purchases_summary['Total Purchase Value'] = gender_purchases_summary['Total Purchase Value'].map('${0:,.2f}'.format)
gender_purchases_summary['Avg Total Purchase per Person'] = gender_purchases_summary['Avg Total Purchase per Person'].map('${0:,.2f}'.format)

gender_purchases_summary #note that the output has male first whereas sample output in the instructions has female first

Purchase Count Average Purchase Price  \
Male                              652                  $3.02   
Female                            113                  $3.20   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Male                             $1,967.64                         $4.07  
Female                             $361.94                         $4.47  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [269]:
purchase_data['Age'].max() #checking max age shows 45 as the highest
bins = [0,9.9,14.9,19.9,24.9,29.9,34.9,39.9,50]
bin_names = ['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']
player_info = players_df.loc[:,['SN','Age']] #players_df is our dataframe for unique players
player_info['Age Range'] = pd.cut(player_info['Age'],bins,labels=bin_names)
age_demo_total = player_info['Age Range'].value_counts()
age_demo_percents = age_demo_total/players

#create and format summary dataframe
age_demo_df = pd.DataFrame({'Total Count':age_demo_total,
                           'Percentage of Players':age_demo_percents},
                          index=bin_names)
age_demo_df['Percentage of Players'] = age_demo_df['Percentage of Players'].map('{0:,.2%}'.format)
age_demo_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [270]:
#we will use the same bins as the above section
player_demo = purchase_data.loc[:,['SN','Age','Price']] #players_df is our dataframe for unique players
player_demo['Age Range'] = pd.cut(player_demo['Age'],bins,labels=bin_names)
purchase_analysis_count = player_demo['Age Range'].value_counts()
#groupby seems to produce a dataframe, so gonna try without specifying a new df 
purchase_analysis_mean = player_demo[['Age Range','Price']].groupby('Age Range')['Price'].mean()
purchase_analysis_total = player_demo[['Age Range','Price']].groupby('Age Range')['Price'].sum()
purchase_analysis_average_total = purchase_analysis_total/age_demo_total #age demo total is the total for unique players

#create and format summary dataframe
age_analysis_df = pd.DataFrame({'Purchase Count':purchase_analysis_count,
                               'Average Purchase Price':purchase_analysis_mean,
                               'Total Purchase Value':purchase_analysis_total,
                               'Avg Total Purchase per Person':purchase_analysis_average_total},
                              index=bin_names)
age_analysis_df['Average Purchase Price'] = age_analysis_df['Average Purchase Price'].map('${0:,.2f}'.format)
age_analysis_df['Total Purchase Value'] = age_analysis_df['Total Purchase Value'].map('${0:,.2f}'.format)
age_analysis_df['Avg Total Purchase per Person'] = age_analysis_df['Avg Total Purchase per Person'].map('${0:,.2f}'.format)
age_analysis_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [271]:
#we can do this directly off the purchase_data dataframe
sn_count = purchase_data.groupby(['SN']).count()['Purchase ID']
sn_average_purchase = purchase_data.groupby(['SN'])['Price'].mean()
sn_total_purchase = purchase_data.groupby(['SN'])['Price'].sum()

#create and format summary dataframe
top_spenders_df = pd.DataFrame({'Purchase Count':sn_count,
                               'Average Purchase Price':sn_average_purchase,
                               'Total Purchase Value':sn_total_purchase})
#pretty weird thing - if I do the formatting before the sort, the order gets completely messed up
top_spenders_df = top_spenders_df.sort_values(['Total Purchase Value'],ascending=False)
top_spenders_df['Average Purchase Price'] = top_spenders_df['Average Purchase Price'].map('${0:,.2f}'.format)
top_spenders_df['Total Purchase Value'] = top_spenders_df['Total Purchase Value'].map('${0:,.2f}'.format)
top_spenders_df.head(5)

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [272]:
#we can do this directly off the purchase_data dataframe
#output might look a little cleaner if we use a grouped dataframe and specify the index in the pd.DataFrame part
item_count = purchase_data.groupby(['Item ID']).count()['Purchase ID']
item_price = purchase_data.groupby(['Item ID'])['Price'].mean() 
item_total_purchase = purchase_data.groupby(['Item ID'])['Price'].sum()

#create and format summary dataframe
popular_items_df = pd.DataFrame({'Purchase Count':item_count,
                               'Item Price':item_price,
                               'Total Purchase Value':item_total_purchase})
popular_items_sort = popular_items_df.sort_values(['Purchase Count'],ascending=False)
popular_items_sort['Item Price'] = popular_items_sort['Item Price'].map('${0:,.2f}'.format)
popular_items_sort['Total Purchase Value'] = popular_items_sort['Total Purchase Value'].map('${0:,.2f}'.format)
popular_items_sort.head(5)

Purchase Count Item Price Total Purchase Value
Item ID                                                
92                   13      $4.61               $59.99
178                  12      $4.23               $50.76
145                   9      $4.58               $41.22
132                   9      $3.22               $28.99
108                   9      $3.53               $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [273]:
profitable_items_sort = popular_items_df.sort_values(['Total Purchase Value'],ascending=False)
profitable_items_sort['Item Price'] = profitable_items_sort['Item Price'].map('${0:,.2f}'.format)
profitable_items_sort['Total Purchase Value'] = profitable_items_sort['Total Purchase Value'].map('${0:,.2f}'.format)
profitable_items_sort.head(5)

Purchase Count Item Price Total Purchase Value
Item ID                                                
92                   13      $4.61               $59.99
178                  12      $4.23               $50.76
82                    9      $4.90               $44.10
145                   9      $4.58               $41.22
103                   8      $4.35               $34.80

## Conclusions Based on Data
###### 1. The game seems to be most popular amongst a younger demographic - notably 44.79% of the playerbase being between the ages of 20 and 24. If we want to expand the range a little bit... 76.74% of the playerbase is between the ages of 15 and 29. It is useful to know what demographic enjoys the game so the developers/company know what sort of content to provide and how to market to their playerbase.
###### 2. Piggybacking on the previous observation, it is noteworthy that the greatest spend among the playerbase is in this group. It would require a little more digging to see if a greater percentage of the players in that age group spent money on items, but is worth doing so to plan future strategies.
###### 3. The items "Final Critic" and "Oathbreaker, Last Hope of the Breaking Storm" (ID 92 and 178 respectively see to be the hottest items amongst those purchased by players. It is worthwhile to look into those items and see what sort of appearance and strength appeals to the playerbase.